File: classes/game/item/CrateItem.cpp
#include "CrateItem.h"
#include "../Match.h"
#include "../Animation.h"
#include "../ImageLibrary.h"
#include "../ParticleLibrary.h"
#include "../Extension.h"
#include "global/SoundManager.h"


CrateItem::CrateItem() {
}

CrateItem::~CrateItem() {
}

void CrateItem::SetupItem() {
	ImageLibrary::CopySprite(this->CurrentSpriteRenderer, IMGLIB->CrateLayerSprites[LayerCount-1]);
	//cocos2d::Rect newrect = CurrentSpriteRenderer->getTextureRect();
	//newrect.origin.y -= 15.0f; newrect.size.height += 15.0f;
	//CurrentSpriteRenderer->setTextureRect(newrect);
}

void CrateItem::RemoveSelf() {
	this->OnBeforeRemove();
	//base.get_gameObject().Recycle();
	this->CurrentSpriteRenderer->removeFromParentAndCleanup(true);
	//delete this;
}

void CrateItem::PlayExplodeAudio() {
	//SoundManager::playEffectSound(SFX::S12); //AudioManager.Play(AudioTag.ItemCrate1, PlayMode.Frame, null, 1f);
}

bool CrateItem::TryExplode() {
	if (this->LatestExplode == Ext_getCurrentFrame())//(long)Time.get_frameCount())
		return false;

	this->LatestExplode = Ext_getCurrentFrame();//(long)Time.get_frameCount();
	if (this->LayerCount == 1)
		return true;

	this->LayerCount--;
	this->StartTranformationToNewLayer();
	return false;
}

///protected:
void CrateItem::PlayOnExplodeStartedAnimation() {
	/*if (ParticlePool.Instance != null) {
		ParticlePool.Instance.PlayParticle(ParticlePool.Instance.CrateExploded, base.get_transform().get_position(), true);
	}//*/
	PTCLLIB->explodeCrate(this->CurrentSpriteRenderer->getPosition());
}

void CrateItem::StartTranformationToNewLayer() {
	int layerCount = this->LayerCount;
	if (layerCount == 3 || layerCount == 2 || layerCount == 1) 	{
		//SoundManager::playEffectSound(SFX::S12); //AudioManager.Play(AudioTag.ItemCrate23, PlayMode.Frame, null, 1f);
		PTCLLIB->explodeCrate(this->CurrentSpriteRenderer->getPosition()); //ParticlePool.Instance.PlayParticle(ParticlePool.Instance.CrateLayer2Or3Exploded, base.get_transform().get_position(), true);
	}
	if(this->LayerCount - 1 < (int)IMGLIB->CrateLayerSprites.size())
		SetupItem();
}

File: doc/porting/CrateItem.cs
using Assets.Scripts.Utils;
using System;
using UnityEngine;

namespace Assets.Scripts.GamePlayScene.Mechanics.Items
{
	public class CrateItem : SpriteBasedItem
	{
		[HideInInspector]
		public int LayerCount;

		public override bool CanFall()
		{
			return false;
		}

		public override bool CanBeTapped()
		{
			return false;
		}

		public override bool CanShadowDrop()
		{
			return true;
		}

		public override bool CanExplodeDueToExplodeInNeigbour(MatchType sourceType)
		{
			return true;
		}

		public override int GetScore()
		{
			return 500;
		}

		public override ItemType GetItemType()
		{
			return ItemType.Crate;
		}

		public override void RemoveAllLayers()
		{
			this.LayerCount = 1;
		}

		public override void PlayExplodeAudio()
		{
			AudioManager.Play(AudioTag.ItemCrate1, PlayMode.Frame, null, 1f);
		}

		public override bool TryExplode()
		{
			if (this.LatestExplode == (long)Time.get_frameCount())
			{
				return false;
			}
			this.LatestExplode = (long)Time.get_frameCount();
			if (this.LayerCount == 1)
			{
				return true;
			}
			this.LayerCount--;
			this.StartTranformationToNewLayer();
			return false;
		}

		protected override void PlayOnExplodeStartedAnimation()
		{
			if (ParticlePool.Instance != null)
			{
				ParticlePool.Instance.PlayParticle(ParticlePool.Instance.CrateExploded, base.get_transform().get_position(), true);
			}
		}

		protected virtual void StartTranformationToNewLayer()
		{
			int layerCount = this.LayerCount;
			if (layerCount == 3 || layerCount == 2 || layerCount == 1)
			{
				AudioManager.Play(AudioTag.ItemCrate23, PlayMode.Frame, null, 1f);
				ParticlePool.Instance.PlayParticle(ParticlePool.Instance.CrateLayer2Or3Exploded, base.get_transform().get_position(), true);
			}
			int num = this.LayerCount - 1;
			if (num < ImageLibrary.Instance.CrateLayerSprites.Length)
			{
				this.CurrentSpriteRenderer.set_sprite(ImageLibrary.Instance.CrateLayerSprites[num]);
			}
		}
	}
}
